### PR TITLE
BUG: infer_dtype with decimal/complex

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [ ] closes #xxxx
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

Does this need a whatsnew?

### CODE REVIEW

File: pandas/_libs/lib.pyx [hunk 1]
```diff
@@ -2546,8 +2565,6 @@ def fast_multiget(dict mapping, ndarray keys, default=np.nan):
         # kludge, for Series
         return np.empty(0, dtype='f8')
 
-    keys = getattr(keys, 'values', keys)
-
```
Comment (jbrockmendel) [inline pandas/_libs/lib.pyx:2550]: unrelated cleanup

File: pandas/_libs/lib.pyx [hunk 2]
```diff
@@ -1414,10 +1414,12 @@ def infer_dtype(value: object, skipna: bool = True) -> str:
             return "time"
 
     elif is_decimal(val):
-        return "decimal"
+        if all(is_decimal(x) for x in values):
```
Comment (jreback) [inline pandas/_libs/lib.pyx:1417]: can you create is_decimal_array to have the same style

Comment (jreback) [reply pandas/_libs/lib.pyx:1417]: i think you actually want to use the validator e.g as a raw NaN would not work here

File: pandas/_libs/lib.pyx [hunk 3]
```diff
@@ -1702,6 +1704,23 @@ cpdef bint is_float_array(ndarray values):
     return validator.validate(values)
 
 
+cdef class ComplexValidator(Validator):
```
Comment (jreback) [inline pandas/_libs/lib.pyx:1707]: is this code hit? don't need new tests?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/_libs/lib.pyx (hunk 2 above)
Comment (jbrockmendel) [reply pandas/_libs/lib.pyx:1417]: can you expand on this?  do you expect something other than Decimal objects to work here?

File: pandas/_libs/lib.pyx (hunk 3 above)
Comment (jbrockmendel) [reply pandas/_libs/lib.pyx:1707]: this is hit by added test case in test_inference

File: pandas/_libs/lib.pyx (hunk 2 above)
Comment (jreback) [reply pandas/_libs/lib.pyx]: NaNs are valid as well here; but more important is that every validator has the same pattern. a is_* function e.g. is_decimal_array, which should call here.

Comment (jbrockmendel) [reply pandas/_libs/lib.pyx]: updated to use the is_* pattern.

since decimal has Decimal("nan"), i wouldnt expect to accept np.nan here

Comment (jreback) [reply pandas/_libs/lib.pyx]: > since decimal has Decimal("nan"), i wouldnt expect to accept np.nan here

yeah ok, though i think we could actually accept np.nan as well and infer a decimal if that is the case (as we do this for all other dtypes). can you open an issue.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks
